application-analytics-default/src/main/java/com/xwiki/analytics/internal/configuration/MostViewedJsonNormaliser.java

```diff
@@ -0,0 +1,152 @@
+     * @return the normalised json as a string.
+     */
+    public JsonNode normaliseData(String jsonString)
+        throws JsonProcessingException, MalformedURLException, UnsupportedResourceReferenceException,
+        CreateResourceTypeException, CreateResourceReferenceException
+    {
+        JsonNode jsonNode = OBJECT_MAPPER.readTree(jsonString);
+        if (jsonNode.isArray()) {
+            processArrayNode(jsonNode);
+        } else {
+            jsonNode = processObjectNode(jsonNode);
+        }
+        return jsonNode;
+    }
+
+    private void processArrayNode(JsonNode jsonNode)
+        throws MalformedURLException, UnsupportedResourceReferenceException, CreateResourceTypeException,
+        CreateResourceReferenceException
+    {
+        for (JsonNode objNode : jsonNode) {
+            if (objNode.isObject()) {
+                ((ObjectNode) objNode).put(DATE, "");
+                if (objNode.has(URL)) {
+                    EntityResourceReference entityResourceReference =
+                        (EntityResourceReference) this.getReference(objNode.get(URL).asText());
+                    ((ObjectNode) objNode).put(LABEL,
+                        entityResourceReference.getEntityReference().getName());
+                    ((ObjectNode) objNode).put(DOCUMENT_REFERENCE,
+                        entityResourceReference.getEntityReference().toString());
+                }
+            }
+        }
+    }
+
+    private ArrayNode processObjectNode(JsonNode jsonNode)
+        throws MalformedURLException, UnsupportedResourceReferenceException, CreateResourceTypeException,
+        CreateResourceReferenceException
+    {
+        ArrayNode arrayNode = OBJECT_MAPPER.createArrayNode();
+        Iterator<String> fieldNames = jsonNode.fieldNames();
+        while (fieldNames.hasNext()) {
+            String date = fieldNames.next();
+            JsonNode childNode = jsonNode.get(date);
+            for (JsonNode objNode : childNode) {
+                if (objNode.isObject()) {
+                    ((ObjectNode) objNode).put(DATE, date);
+                    if (objNode.has(URL)) {
+                        EntityResourceReference entityResourceReference =
+                            (EntityResourceReference) this.getReference(objNode.get(URL).asText());
+                        ((ObjectNode) objNode).put(LABEL,
+                            entityResourceReference.getEntityReference().getName());
+                        arrayNode.add(objNode);
+                        ((ObjectNode) objNode).put(DOCUMENT_REFERENCE,
+                            entityResourceReference.getEntityReference().toString());
+                    }
+                }
+            }
+        }
+        return arrayNode;
+    }
+
+    private ResourceReference getReference(String url)
+        throws MalformedURLException, CreateResourceTypeException, UnsupportedResourceReferenceException,
+        CreateResourceReferenceException
```

Comment:
You should just catch these Exceptions here, so you can maybe return a null value in case something happened (an also see when you call this method what to do if you receive a null value)